src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs

```diff
@@ -209,19 +210,69 @@ public RequestDelegate CreateMiddleware(RequestDelegate next)
             }
         }
 
+        // Performance optimization: Precompute and cache the results for each parameter
+        var parameterData = new (bool hasServiceKey, object? key, Type parameterType, Type declaringType)[parameters.Length];
```

Comment:
You only really need to store an array of `object? key`. The other properties are already being captured by the closure and the check for whether to get a keyed service is `key == null ? GetService : GetKeyedService`.
https://github.com/dotnet/runtime/blob/26bcc1d12f895d654e24f7cbcfed466b6a34e65f/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L381

@benjaminpetit should we be throwing if `null` is passed in for key in `FromKeyedServices`?

Reply:
@BrennanConroy, I'll change the logic according to your advice.

The reason as to why I've precomputed and cached the parameter- and declaringType is due to the fact that these are actually method calls to get the property value – twice since we'd be retrieving two values. As well as obtain the parameter from the parameters array. Since with the tuple we'd have direct memory access, I figured it would be performing better during high throughput applications.

Reply:
I love improving perf, so appreciate you thinking about that 😃 
But we probably shouldn't prematurely optimize without measuring. You can try it out in https://github.com/dotnet/aspnetcore/tree/main/src/Http/Http.Abstractions/perf/Microbenchmarks if you're up for it.

Reply:
@BrennanConroy, totally up for it! Tomorrow though, I need some sleep first 🥱

Reply:
@BrennanConroy, I've added some benchmarks, as well as the possible optimizations. It seems that precomputing and caching all the parameters is the most optimized option. 
[benchmarks.txt](https://github.com/dotnet/aspnetcore/files/15395939/benchmarks.txt)

I've also introduced a new optimization which would cache a delegate based on the GetService or GetKeyedService method call, however, calling to the cached delegate itself seems to have more overhead than the key check and direct access to the method.

Based on these benchmarks, should we go with the optimization to cache all the parameters rather than just the key? What's your opinion on the matter?

Reply:
@BrennanConroy I'm wondering about your follow up on the above.